### PR TITLE
S3 disk unstable reads test

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by excitoon** [comment]
Progress on #22988, #23976, #24311, #24885 and #26249

**Comment by excitoon** [comment]
This is a strange fail which only appeared under ASAN without code changes.

**Comment by excitoon** [comment]
I can not reproduce it locally.

**Comment by excitoon** [comment]
Trying to revert 3d80608deb6a991d505569cdd9ecb8d05cfb6c44

**Comment by excitoon** [comment]
Looks like there is some fresh bug in the following commit range, because I could not reproduce it before rebase: c18c16855a6f8ce3f655fdda0ac3c6d60ecf7d04..14ef66e49b62e981d2011e924f687589e3cf81eb

**Comment by excitoon** [comment]
d63a5e1c9651a2b74a09cb4cd166241b2f2e9426:

```
E           Code: 32. DB::Exception: Received from 172.16.1.6:9000. DB::Exception: Attempt to read after eof: (while reading column id): (while reading from part /var/lib/clickhouse/disks/unstable_s3/store/971/971a795c-624e-4079-971a-795c624ee079/20210809_3_3_0/ from mark 1522 with max_rows_to_read = 65505): While executing MergeTreeThread. Stack trace:
```

File: tests/integration/test_merge_tree_s3/s3_mocks/unstable_proxy.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    def do_HEAD(self):
+        content_length = self.headers.get("Content-Length")
+        data = self.rfile.read(int(content_length)) if content_length else None
+        r = request(self.command, f"http://{UPSTREAM_HOST}{self.path}", headers=self.headers, data=data)
+        self.send_response(r.status_code)
+        for k, v in r.headers.items():
+            self.send_header(k, v)
+        self.end_headers()
+        if random.random() < 0.25 and len(r.content) > 1024*1024:
+            r.content = r.content[:len(r.content)//2]
```
Comment (Jokser) [inline tests/integration/test_merge_tree_s3/s3_mocks/unstable_proxy.py:54]: Could you please describe in more details: when we send only half of the content, what happens on CH side?

Comment (excitoon) [reply tests/integration/test_merge_tree_s3/s3_mocks/unstable_proxy.py:54]: Just a way to simulate connection drop, at the middle of request :man_shrugging: . On CH side it shall understand that not all data is received and retry.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by excitoon** [comment]
This PR can not affect performance.

**Comment by alexey-milovidov** [comment]
Performance test was broken in #27492.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**